src/main/flight/position.c

```diff
@@ -80,152 +79,154 @@ PG_REGISTER_WITH_RESET_TEMPLATE(positionConfig_t, positionConfig, PG_POSITION, 4
 
 PG_RESET_TEMPLATE(positionConfig_t, positionConfig,
     .altitude_source = DEFAULT,
-    .altitude_prefer_baro = 100,
+    .altitude_prefer_baro = 100, // percentage 'trust' of baro data
     .altitude_lpf = 300,
     .altitude_d_lpf = 100,
 );
 
 #if defined(USE_BARO) || defined(USE_GPS)
-static bool altitudeOffsetSetBaro = false;
-static bool altitudeOffsetSetGPS = false;
+static bool gpsAltitudeZeroed = false;
+static bool okToZeroGps = false;
 
 void calculateEstimatedAltitude()
 {
-    float baroAltCm = 0;
-    float gpsAltCm = 0;
-    float altitudeToFilterCm = 0.0f;
-
-    float gpsTrust = 0.3; //conservative default
-    bool haveBaroAlt = false;
-    bool haveGpsAlt = false;
-
-// *** Set baroAlt once its calibration is complete (cannot arm until it is)
+    float baroAltCm = 0.0f;
+    static float gpsAltCm = 0.0f; // to hold last value if transient loss of 3D fix
+    float gpsTrust = 0.3f; // conservative default
+    bool haveBaroAlt = false; // true if baro exists and has been calibrated on power up
+    bool haveGpsAlt = false; // true if GPS is connected and while it has a 3D fix
+    bool baroAltitudeZeroed = false; // true if zeroed on arming, and armed
+
+    // *** Set baroAlt once its calibration is complete (cannot arm until it is)
 #ifdef USE_BARO
     if (sensors(SENSOR_BARO)) {
         baroAltCm = getBaroAltitude();
         if (baroIsCalibrated()) {
             haveBaroAlt = true;
         }
-    }
-
-    // *** Zero Baro Altitude on arming (every time we re-arm, also)
+    // *** Zero Baro Altitude on arming
     // note that arming is blocked until baro 'calibration' (baro ground zeroing) is complete, no need to check status of haveBaroAlt
-    // this code adds a secondary zeroing to whatever baro altitude value exists on arming
     // since props spin on arming, we want the last value before arming
-    // provided that this function is called before significant motor spin-up has occured, this may not be a big problem
-
-    static float baroAltOffsetCm = 0;
-    if (ARMING_FLAG(ARMED)) {
-        baroAltCm -= baroAltOffsetCm; // use the last offset value from the disarm period
-        altitudeOffsetSetBaro = true; // inevitable, but needed if no GPS
-    } else {
-        baroAltOffsetCm = baroAltCm; // while disarmed, keep capturing current altitude to zero any offset once armed
+        static float baroAltOffsetCm = 0.0f;
```

Comment:
Thanks.
I think @KarateBrot and @Pichimeter will rewrite the sensor integration code completely in future, and that would change how and where filtering will be applied to the various input streams.
Filtering baro continuously would make sense since the filtered value would be useful for zeroing on arming.
However it seemed to me to be more efficient to filter only the output of the mixed baro+gps signal, rather than filtering baro and then also needing to filter GPS independently.
For now let's leave it like this for merging; it works OK and will be changed for sure in the near  future.

Reply:
> However it seemed to me to be more efficient to filter only the output of the mixed baro+gps signal, rather than filtering baro and then also needing to filter GPS independently.

It's more efficient but you also will lose some bandwidth of the baro signal because you need to filter for the lower GPS rate. This is a bottleneck for the baro components in the mixed signal. But I'll go over that and other parts of the code anyways starting next week when I have more free time. 👍